### ISSUE TITLE
I2I: a method to unblock AMP contributors in the face of flaky tests on Travis

### ISSUE THREAD

**Issue by danielrozenberg** [ampproject/amphtml#20744]
## Background

The ampproject/amphtml GitHub+Travis project is suffering from a long standing problem of flaky tests that hinder developer velocity, sometimes for days, and reduces developer confidence in our tooling.

The root cause stems for this flakiness problem stems from a combination of factors:

1.  Our testing infrastructure is complex: we run unit tests, integration tests, end-to-end tests, visual diff tests, and SauceLabs tests
1.  Some tests are outdated: the oldest test file was last updated in November 2015 (34 tests were last updated at or before 2016; 218 tests at or before 2017); outdated test files means that they were written before some testing platform changes, which might make them incompatible
1.  SauceLabs proxy often fails to open a tunnel
1.  SauceLabs tests often break when new versions of browsers are promoted from beta to stable

We attempted several solutions to this problem, including early warning for tests that break in new browser versions (factor 4), retrying the SauceLabs tunnel several times (factor 3), encouraging more structured test writing (tangentially related to factor 2).

We also intend to split our test execution to several smaller Travis shards, so that each can be re-run individually (factor 1). That is, we would execute multiple parallel shards of build/test tasks, each shard running inside larger, linear stages (with data persistence between them). A description of these other changes will arrive in a later I2I issue. The work described in this I2I can be performed in parallel to the sharding effort.

## Goals

This I2I proposes a solution that would allow the weekly build cop to override a failing test, in effect allowing us to ignore flaky tests. While deciding on the solution to the problem, we wanted to priorities the following:

1.  Increase developer velocity and confidence in our tooling
1.  Ensure that the build cop has the final say in unblocking a PR due to flaky tests
1.  Avoid permanently marking tests as flaky with a "set it and forget it" mentality, which will reduce our total test coverage and confidence in the testing platform

### Non-goals

This effort does not involve any changes to the testing platform or to the tests themselves. There are other efforts we intend to pursue to do both. The intent of this effort is to unblock developers and free them from this recurring issue, while still pursuing a longer-term effort to improve the testing platforms and tests.

## Overview

![image](https://user-images.githubusercontent.com/1839738/52507523-cb3c8600-2bbf-11e9-868d-e90f04cb6cf9.png)

Similar to the [Bundle Size GitHub App](https://github.com/ampproject/amphtml/issues/19146), we will create an AMP Tests GitHub App that will block a PR if any of the test platforms reports failure, until manually approved.

The app will be written using [Probot](https://github.com/probot/probot) and will run on AppEngine. State changes will be driven by calls from the Travis build into the app's API. Travis shard restarts will replace and supercede previous status check lines.

![image](https://user-images.githubusercontent.com/1839738/52507711-57e74400-2bc0-11e9-9345-5f0783ddbc28.png)

The App itself will listen to webhook requests from the GitHub API to invalidate results when a PR is synchronized (new commits are pushed/force-pushed), and will listen to requests when a PR is created to store credential so that it can perform actions on GitHub.

When a test fails and its status check's state is set to `Action Required`, the `Details` link should point to a page with details of the failures, and for the Approvers a link to approve skipping those tests, with a confirmation screen to prevent accidental skips.

Tests on Travis should be modified to not fail the build on PRs (they should still fail on push builds) since this check will replace the existing failure condition.

Questions/comments left:

*   When run is finished, get JSON output from Travis from [Raghu's PR](https://github.com/ampproject/amphtml/pull/20590) to determine result
*   How to assign current build cop the correct permissions to approve automatically? Currently the build cop rotation is Google-internal. As we begin to add non-googlers to the rotation, a non-Google system will need to be used

### Status message

The status messages on the app should be concise, since there is not a lot of space in the UI

*   :x:	**ampproject/tests (integration)** — _Action Required — FAILED: 4 tests | PASSED: 928 tests_
*    :small_orange_diamond:	**ampproject/tests (e2e)** — _Queued — e2e tests have not yet started_
*    :small_orange_diamond:	**ampproject/tests (sauce-labs)** — _Pending — sauce-labs tests are running_
*   :heavy_check_mark:	**ampproject/tests (unit)** — _PASSED_: 1872 tests
*   :heavy_check_mark:	**ampproject/tests (integration)** — PASSED: **RESOLVED BY BUILD COP** | _FAILED_: 4 tests | _PASSED_: 928 tests
*   :black_small_square:	**ampproject/tests (e2e)** — Skipped — e2e tests not required

### Details page

The Details page for each Failed/Passed status (including Passed due to Approver skipping the tests) should list the failing tests as they would be displayed in the Travis logs. An approved skip test should list the GitHub username of the Approver.

The Details page for Queued/Pending/Skipped status should contain a short blurb about the situation, and a link to the documentation on testing.

### Documentation

Let's not forget to update the documentation on testing ;)

/cc @ampproject/wg-approvers @ampproject/wg-infra

**Comment by danielrozenberg** [comment]
This is no longer as big an issue as it used to be when this I2I was added. The test-status bot exposes the status of the various tests but does not let the buildcop actually bypass the Travis checks. I doubt we'll ever get to that point again, so closing this. Feel free to reopen if you think we should do this
